### ISSUE TITLE
dataset inconsistencies depending on creation method

### ISSUE THREAD

**Issue by ungil** [Microsoft/LightGBM#1101]
I expected the following example to produce the same output when the datasets are created passing the filename to lgb.Dataset and when the files are first read into python.

```
import lightgbm as lgb
import pandas as pd

params = {'task': 'train', 'boosting_type': 'gbdt', 'objective': 'regression', 'metric': 'auc', 'verbose': 0}

print('reading datafiles as panda objects')
df_train = pd.read_csv('~/LightGBM-2.0.11/examples/regression/regression.train', header=None, sep='\t')
df_test = pd.read_csv('~/LightGBM-2.0.11/examples/regression/regression.test', header=None, sep='\t')

y_train = df_train[0].values
y_test = df_test[0].values
X_train = df_train.drop(0, axis=1).values
X_test = df_test.drop(0, axis=1).values

lgb_train = lgb.Dataset(X_train, y_train)
lgb_eval = lgb.Dataset(X_test, y_test, reference=lgb_train)

gbm = lgb.train(params, lgb_train, num_boost_round=5, valid_sets={lgb_train,lgb_eval})

print('reading datafiles directly')
lgb_train2 = lgb.Dataset('~/LightGBM-2.0.11/examples/regression/regression.train')
lgb_eval2 = lgb.Dataset('~/LightGBM-2.0.11/examples/regression/regression.test', reference=lgb_train2)

gbm2 = lgb.train(params, lgb_train2, num_boost_round=5, valid_sets={lgb_train2,lgb_eval2})
```

In my tests, it looks like it's working fine on Ubuntu:
```
reading datafiles as panda objects
[1]	training's auc: 0.768801	valid_1's auc: 0.771535
[2]	training's auc: 0.790216	valid_1's auc: 0.787265
[3]	training's auc: 0.800972	valid_1's auc: 0.784507
[4]	training's auc: 0.805503	valid_1's auc: 0.789836
[5]	training's auc: 0.809702	valid_1's auc: 0.79182
reading datafiles directly
[1]	training's auc: 0.768801	valid_1's auc: 0.771535
[2]	training's auc: 0.790216	valid_1's auc: 0.787265
[3]	training's auc: 0.800972	valid_1's auc: 0.784507
[4]	training's auc: 0.805503	valid_1's auc: 0.789836
[5]	training's auc: 0.809702	valid_1's auc: 0.79182
```

but on Windows passing the filenames to lgb.Dataset gives strange results:
```
reading datafiles as panda objects
[1]	training's auc: 0.768801	valid_0's auc: 0.771535
[2]	training's auc: 0.790216	valid_0's auc: 0.787265
[3]	training's auc: 0.800972	valid_0's auc: 0.784507
[4]	training's auc: 0.805503	valid_0's auc: 0.789836
[5]	training's auc: 0.809702	valid_0's auc: 0.79182
reading datafiles directly
[1]	training's auc: 0.53515	valid_0's auc: 0.529049
[2]	training's auc: 0.554074	valid_0's auc: 0.54619
[3]	training's auc: 0.570398	valid_0's auc: 0.561323
[4]	training's auc: 0.585889	valid_0's auc: 0.575787
[5]	training's auc: 0.599194	valid_0's auc: 0.590106
```

as it does on MacOS (the results are also wrong, but in a slightly different way!):
```
reading datafiles as panda objects
[1]	training's auc: 0.768801	valid_0's auc: 0.771535
[2]	training's auc: 0.790216	valid_0's auc: 0.787265
[3]	training's auc: 0.800972	valid_0's auc: 0.784507
[4]	training's auc: 0.805503	valid_0's auc: 0.789836
[5]	training's auc: 0.809702	valid_0's auc: 0.79182
reading datafiles directly
[1]	training's auc: 0.53515	valid_1's auc: 0.529049
[2]	training's auc: 0.554004	valid_1's auc: 0.546335
[3]	training's auc: 0.570333	valid_1's auc: 0.561113
[4]	training's auc: 0.585754	valid_1's auc: 0.575142
[5]	training's auc: 0.599043	valid_1's auc: 0.589348
```

Note as well that the name of the validation set is not consistent (it can be "valid_1" or "valid_0", it may change with repeated executions of the script).

The problem is not restricted to the latest version of the module, the same output is obtained since 2.0.7 (modules installed using pip).

I found this example investigating an issue I found with the C API interface, so I suspect this is not really a problem with the Python module but a manifestation of something deeper.

**Comment by guolinke** [comment]
@ungil 
there are ".init" files in that data folder. As a result, these "init" file will be loaded automatically. 
You can remove them and test again.

**Comment by guolinke** [comment]
you also can force set the init_score to zero.

```
lgb_train2 = lgb.Dataset('examples/regression/regression.train')
lgb_eval2 = lgb.Dataset('examples/regression/regression.test', reference=lgb_train2)
lgb_train2.construct()
lgb_eval2.construct()

lgb_train2.set_init_score([0 for i in range(lgb_train2.num_data())])
lgb_eval2.set_init_score([0 for i in range(lgb_eval2.num_data())])
```

I am not sure how the result in  Ubuntu is normal, maybe you didn't have "init" file in that test?

**Comment by ungil** [comment]
Thank you for your quick reply! I'm happy to learn that there is a simple explanation that I was missing. And you're right, on Ubuntu I was taking the datafiles from an old version without .init files (the output with .init files is identical to the MacOS output).

**Comment by ungil** [comment]
@guolinke your suggestion of setting init_score to zero works only in part. If I include two metrics
```
params = {'task': 'train', 'boosting_type': 'gbdt', 'objective': 'regression', 'metric': {'auc','l2'}, 'verbose': 0}
```
AUC gives the same results but L2 doesn't:
```
reading datafiles as panda objects
[1]	training's l2: 0.238643	training's auc: 0.768801	valid_0's l2: 0.237569	valid_0's auc: 0.771535
[2]	training's l2: 0.229898	training's auc: 0.790216	valid_0's l2: 0.229057	valid_0's auc: 0.787265
[3]	training's l2: 0.222344	training's auc: 0.800972	valid_0's l2: 0.222616	valid_0's auc: 0.784507
[4]	training's l2: 0.215907	training's auc: 0.805503	valid_0's l2: 0.216681	valid_0's auc: 0.789836
[5]	training's l2: 0.210251	training's auc: 0.809702	valid_0's l2: 0.212015	valid_0's auc: 0.79182
reading datafiles directly
[LightGBM] [Info] Loading initial scores...
[LightGBM] [Info] Loading initial scores...
[1]	training's l2: 0.466908	training's auc: 0.768801	valid_1's l2: 0.480395	valid_1's auc: 0.771535
[2]	training's l2: 0.414793	training's auc: 0.790216	valid_1's l2: 0.427781	valid_1's auc: 0.787265
[3]	training's l2: 0.372109	training's auc: 0.800972	valid_1's l2: 0.386183	valid_1's auc: 0.784507
[4]	training's l2: 0.337216	training's auc: 0.805503	valid_1's l2: 0.351224	valid_1's auc: 0.789836
[5]	training's l2: 0.308512	training's auc: 0.809702	valid_1's l2: 0.322429	valid_1's auc: 0.79182
```
Is this behaviour expected? I may be overlooking something obvious again...

**Comment by guolinke** [comment]
@ungil you can set "boost_from_average=False". 
"boost_from_average" will be disable when has init_score, even when init_score are all zeros.
